### ISSUE TITLE
search: prevent OOMs on index search against thousands of repos in a sourcegraph docker container

### ISSUE THREAD

**Issue by ijt** [sourcegraph/sourcegraph#3057]
We have received reports that if indexed search is enabled after thousands of repos are already cloned, then the instance becomes unusable. The goal is to make it possible to safely enable indexing on instances that already have thousands of cloned repos.

See https://github.com/sourcegraph/sourcegraph/issues/2176#issuecomment-475428889 for background.

Customers (not exhaustive):

- https://app.hubspot.com/contacts/2762526/company/887065580

**Comment by slimsag** [comment]
I took a look at the `sourcegraph-zoekt-indexserver` code. It is actually really sane and does all work consecutively on 0.25 of the number of cores the machine has. On the system reported in #2970 with 8 cores, it would at max consume 2 cores for indexing it seems.

https://github.com/sourcegraph/zoekt/blob/8aa57bd35909938707f0af2bfaeaf24acec721b3/cmd/zoekt-sourcegraph-indexserver/main.go#L150

https://github.com/sourcegraph/zoekt/blob/8aa57bd35909938707f0af2bfaeaf24acec721b3/cmd/zoekt-sourcegraph-indexserver/main.go#L160

there is one exception, which is that it also does indexing when `/` has a request made to it, but it doesn't seem we use that?

**Comment by keegancsmith** [comment]
Why is this a new issue? I would suggest we avoid fragmenting the discussion and stick to unless I am mistaken on the difference between the two?https://github.com/sourcegraph/sourcegraph/issues/2176#issuecomment-475428889

**Comment by ijt** [comment]
As far as I can tell, enabling indexed search by default (#2176) is a distinct issue from making it safe to turn it on after lots of repos are added (#3057). I think the comment you linked to would make more sense on this issue.

**Comment by ijt** [comment]
Replying to it here:

> > Why would zoekt-sourcegraph-indexserver have a hard time if indexing is enabled after thousands of repos have been added to an instance?
> 
> zoekt-sourcegraph-indexserver is bounded in the amount of resources used. In particular it has a tuneable maximum amount index jobs it runs at one time (related to the number of cores). Each job tries to use as much CPU as possible, but is bounded in the amount of RAM it will use. Essentially it will shard a repo into 100mb chunks, where 100mb is the output. So it should use roughly 100mb * jobs when indexing at capacity. Note: I may be remembering the number incorrectly for maximum shard size.
> 
> The webserver is problematic after indexing is complete because it is bounded by the size of the working copy of all cloned repositories. It keeps in memory the index (trigram -> posting lists), and it builds result sets in memory (results sets are bounded though). I can't remember what the usual factor is offhand, but the number in my head is 0.7 the size of the working copy. Note: it excludes a lot of files like binary files / large generated files / etc so the working copy size can actually be quite reasonable compared to what du -h tells you locally.

Can you link to the definition of the variable where the in-memory index is stored?

Do you think it's possible to reduce the memory consumption to whatever it would be for a single repo?

**Comment by ijt** [comment]
Interesting idea from @ggilmore : maybe the problem is that there is some big repository with so many terms to index that it causes the indexer to exhaust the system's memory. Trying this now with `github.com/torvalds/linux`.

**Comment by ijt** [comment]
We just tried cloning `github.com/torvalds/linux` (3.4G) and then indexing in a container with mem limit of 2G and it worked fine.

**Comment by keegancsmith** [comment]
Just a technicality: Zoekt doesn't index terms, it indexes trigrams. In fact for a large repository it probably does quite well because those will have larger posting lists => better chance for compression.

Here is the data held in memory: https://sourcegraph.com/github.com/google/zoekt/-/blob/indexdata.go#L28

Note there is a memoryUse function which calculates how much memory is used. For context I mentioned in my previous comment shards. Each shard is an indexData instance.

Also note that zoekt already uses mmap for `IndexFile` the first field in the above struct. That is contains filepaths and file contents which the indexes point into (generally when you see a reference to an offset, its an offset in that file)

**Comment by ijt** [comment]
Based on this, I'm guessing we can trigger the issue by adding a repo containing a big file containing nothing but 'aaaa...', indexing it, and then searching for 'aaa'. I'll try that now.

**Comment by keegancsmith** [comment]
That probably won't work because the file will be ignored due to its size. You could make a big repo which is just the linux repo 10k times. Zoekt doesn't do file dedup so should simulate it.

**Comment by ijt** [comment]
cc @ryan-blunden

**Comment by ijt** [comment]
Here's a way to get indexed search to crash:

On a linux box, launch a sourcegraph docker container with `--memory=700m`.
Configure it to include these repos each containing 500m of files with `aaaaaa` repeated many times:
    github.com/ijt/bigrepo564024188
    github.com/ijt/bigrepo224990488
Enable indexing.
Wait for the repos to be indexed.
Search for the string `aaa`.
The container should crash because of the zoekt-webserver process running out of memory, and the search results may look something like this:

<img width="1440" alt="Screen Shot 2019-04-04 at 4 50 43 PM" src="https://user-images.githubusercontent.com/15530/55595939-1d4ee300-56fb-11e9-9992-d58ba5965ac3.png">

I've also seen it show an error message about and EOF, and at other times a 502 gateway error.

**Comment by ijt** [comment]
Enabling indexed search after thousands of repos are already cloned does appear to be safe. The indexing step tends to be pretty good about not using too much memory. However, once the repos are indexed, particular queries containing trigrams with many hits can cause `zoekt-webserver` to use lots of memory. If that happens, one possibility is to increase the amount of memory available. Another is to run `zoekt-webserver` in a separate container to limit the blast radius if it OOMs. @ggilmore added something for this recently.

**Comment by keegancsmith** [comment]
Was this meant to be closed? It isn't clear to me what was done to make it useable.

**Comment by ijt** [comment]
@keegancsmith, I guess you're pointing to how there is still the problem that once indexing is finished, it's still possible to crash the whole container by issuing a query whose trigrams have large posting lists relative to the amount of memory available to the container. That's still true.

The title is a bit misleading though as it stands, suggesting that the timing of enabling indexed search is somehow important. I can change that and reopen it, but I don't think it makes sense for this to be a high priority right now since we have the mitigation of [using docker compose](https://github.com/sourcegraph/sourcegraph/pull/3332) to keep indexed search OOMs from crashing everything.

**Comment by keegancsmith** [comment]
Yeah if it isn't a high priority then close this out. No point letting it just rot in our issue tracker. Lets just ensure that admins can discover the docker compose approach in our docs if they run into these sort of issues.

**Comment by ijt** [comment]
Okay, re-closing it.
